packages/yew-agent/src/reactor/hooks.rs

```diff
@@ -0,0 +1,279 @@
+/// The callback will be updated on every render to make sure captured values (if any) are up to
+/// date.
+#[hook]
+pub fn use_reactor_bridge<R, F>(on_output: F) -> UseReactorBridgeHandle<R>
+where
+    R: 'static + Reactor,
+    F: Fn(ReactorEvent<R>) + 'static,
+{
+    let ctr = use_reducer(BridgeIdState::default);
+
+    let worker_state = use_context::<ReactorProviderState<R>>()
+        .expect_throw("cannot find a provider for current agent.");
+
+    let on_output = Rc::new(on_output);
+
+    let on_output_ref = {
+        let on_output = on_output.clone();
+        use_mut_ref(move || on_output)
+    };
+
+    // Refresh the callback on every render.
+    {
+        let mut on_output_ref = on_output_ref.borrow_mut();
+        *on_output_ref = on_output;
+    }
+
+    let tx = use_memo((worker_state, ctr.inner), |(state, _ctr)| {
+        let bridge = state.create_bridge();
+
+        let (tx, mut rx) = bridge.split();
+
+        spawn_local(async move {
+            while let Some(m) = rx.next().await {
+                let on_output = on_output_ref.borrow().clone();
+                on_output(ReactorEvent::<R>::Output(m));
+            }
+
+            let on_output = on_output_ref.borrow().clone();
+            on_output(ReactorEvent::<R>::Finished);
+        });
+
+        RwLock::new(tx)
+    });
+
+    UseReactorBridgeHandle {
+        tx: tx.clone(),
+        ctr: ctr.dispatcher(),
+    }
+}
+
+/// Hook handle for the [`use_reactor_subscription`] hook.
+pub struct UseReactorSubscriptionHandle<R>
+where
+    R: 'static + Reactor,
+{
+    bridge: UseReactorBridgeHandle<R>,
+    outputs: Vec<Rc<<R::Scope as ReactorScoped>::Output>>,
+    finished: bool,
+    ctr: usize,
+}
+
+impl<R> UseReactorSubscriptionHandle<R>
+where
+    R: 'static + Reactor,
+{
+    /// Send an input to a reactor agent.
+    pub fn send(&self, msg: <R::Scope as ReactorScoped>::Input) {
+        self.bridge.send(msg);
+    }
+
+    /// Returns whether the current bridge has received a finish message.
+    pub fn finished(&self) -> bool {
+        self.finished
+    }
+
+    /// Reset the subscription.
+    ///
+    /// This disconnects the old bridge and re-connects the agent with a new bridge.
+    /// Existing outputs stored in the subscription will also be cleared.
+    pub fn reset(&self) {
+        self.bridge.reset();
+    }
+}
+
+impl<R> Clone for UseReactorSubscriptionHandle<R>
+where
+    R: 'static + Reactor,
+{
+    fn clone(&self) -> Self {
+        Self {
+            bridge: self.bridge.clone(),
+            outputs: self.outputs.clone(),
+            ctr: self.ctr,
+            finished: self.finished,
+        }
+    }
+}
+
+impl<R> fmt::Debug for UseReactorSubscriptionHandle<R>
+where
+    R: 'static + Reactor,
+    <R::Scope as ReactorScoped>::Input: fmt::Debug,
+    <R::Scope as ReactorScoped>::Output: fmt::Debug,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("UseReactorSubscriptionHandle<_>")
```

Comment:
Same as before; using `type_name` instead of `_` can help readability of debug output